crates/node/src/roles/worker.rs

```diff
@@ -86,73 +85,140 @@ impl WorkerRole {
 
-        let mut worker_handle = tokio::spawn(self.worker.run(inner_shutdown_watch));
+        component_set.spawn(
+            self.worker
+                .run(inner_shutdown_watch)
+                .map_ok(|_| "worker")
+                .map_err(WorkerRoleError::Worker),
+        );
+        component_set.spawn(
+            Self::reload_schemas(
+                subscription_controller,
+                self.schemas,
+                // todo: make this configurable
+                Channel::builder("http://127.0.0.1:5122/".parse().expect("valid uri"))
+                    .connect_lazy(),
+            )
+            .map_ok(|_| "schema-update"),
+        );
 
         tokio::select! {
             _ = shutdown_signal => {
                 info!("Stopping worker role");
-                let _ = tokio::join!(inner_shutdown_signal.drain(), admin_handle, meta_handle, worker_handle, bifrost_handle);
-            },
-            result = &mut meta_handle => {
-                result.map_err(WorkerRoleError::MetaPanic)??;
-                panic!("Unexpected termination of meta.");
+                inner_shutdown_signal.drain().await;
+                // ignoring result because we are shutting down
+                let _ = tokio::join!(component_set.shutdown(), bifrost_join_handle);
             },
-            result = &mut admin_handle => {
-                result.map_err(WorkerRoleError::AdminPanic)??;
-                panic!("Unexpected termination of admin.");
-            },
-            result = &mut worker_handle => {
-                result.map_err(WorkerRoleError::WorkerPanic)??;
-                panic!("Unexpected termination of worker.");
+            Some(component_result) = component_set.join_next() => {
+                let component_name = component_result.map_err(WorkerRoleError::ComponentPanic)??;
+                panic!("Unexpected termination of component '{component_name}'");
+            }
+            bifrost_result = &mut bifrost_join_handle => {
+                bifrost_result.map_err(WorkerRoleError::ComponentPanic)??;
+                panic!("Unexpected termination of bifrost service");
             }
-            result = &mut bifrost_handle => {
-                result.map_err(WorkerRoleError::BifrostPanic)??;
-                panic!("Unexpected termination of bifrost service.");
-            },
         }
 
         Ok(())
     }
+
+    async fn reload_schemas<SC>(
+        subscription_controller: Option<SC>,
+        schemas: Schemas,
+        schema_channel: Channel,
+    ) -> Result<(), WorkerRoleError>
+    where
+        SC: SubscriptionController + Clone + Send + Sync,
+    {
+        // todo: make this configurable
+        let mut fetch_interval = tokio::time::interval(Duration::from_secs(5));
+        let mut schema_grpc_client =
+            restate_node_services::schema::schema_client::SchemaClient::new(schema_channel);
```

Comment:
Same concern/question about getting invocations before the receiving the schema update here in steady state, but this is less concerning in this single-node setup since ingress uses the same schema as worker. The remaining risk is invocation failures in the time gap between registering a new service and ingress/worker being notified (up to 5 seconds)

Reply:
Yes, it can currently happen that for the refresh interval (currently 5s) the ingress will reject invocations for a newly registered service. Possible ways to close the gap/improve the situation could be:

* Ingress/worker trying to refresh its schema information on a miss (probably some rate limiting is still desirable to prevent DoS)
 * Worker can subscribe to schema information changes; Admin role tries to notify workers on schema changes on a best-effort basis.

Reply:
Do you think that a 5s gap (E(T) = 2.5s) between registering a new service and being able to invoke it is a reasonable default? Since every individual fetch request can become somewhat costly (they are fetching all update commands and then re-apply them to create the resulting `Schemas` instance), I tried to balance it a bit.

Reply:
I don't think this is acceptable from a user perspective. One possible path forward is what's suggested in https://github.com/restatedev/restate/pull/1180#discussion_r1489824266 

Alternatively, we can make service update a two-step process. Where you discover a deployment and it's added as inactive first, then the user enables/activates it. This has the benefit of giving users control over shifting traffic from one deployment to the other (we can offer gradual rollout control) but from our perspective, it removes the expectation of atomic schema updates and it'll only be required when the user "activates" the schema.

Reply:
Makes sense. With the eager notifications from the metas to the workers, while not being perfect, it should already close the gap for most cases.